### PR TITLE
Fix crash issue "Cannot read property 'toLowerCasePropertyNames' of undefined"

### PR DESCRIPTION
**Pull request by narek-king**
Related issue link: https://github.com/aws/amazon-chime-sdk-js/issues/1052

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by ltrung** [comment]
@narek-king Thank for fixing this. Do you mind updating the unit test for the case that causes the crash here https://github.com/aws/amazon-chime-sdk-js/blob/master/test/meetingsession/MeetingSessionConfiguration.test.ts ?

**Comment by narek-king** [comment]
@ltrung I fixed the unit tests. The CI fails and I am not sure why. Can you please review it?

**Comment by ltrung** [comment]
> @ltrung I fixed the unit tests. The CI fails and I am not sure why. Can you please review it?

@narek-king  currently forked PR would not be able to trigger our integration test as it cannot access GitHub secret. We are working on fixing it.
Regarding your PR, could you clarify what the issue is? I could run the unit test you added successfully without the fix. [].map should not cause any issue.

**Comment by narek-king** [comment]
> > @ltrung I fixed the unit tests. The CI fails and I am not sure why. Can you please review it?
> 
> @narek-king currently forked PR would not be able to trigger our integration test as it cannot access GitHub secret. We are working on fixing it.
> Regarding your PR, could you clarify what the issue is? I could run the unit test you added successfully without the fix. [].map should not cause any issue.

@ltrung I agree that [].map should not cause any issue but I have a real case that it didn't work. This fixed my problem and I am currently using the SDK from my forked repo. In my opinion, the problem is not related to [].map but more of recursively calling the `toLowerCasePropertyName` method. In the scope, the reference of the method is not accessible. I was thinking to refactor to call as a static method but in this case, it doesn't make sense. 
I was thinking of another case where the input is [[]] but didn't have time to check.
image
I am not sure why this is hapenning^

File: CHANGELOG.md [hunk 1]
```diff
@@ -15,6 +15,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
   `attendeePresenceReceived` meeting event.
 - Add the file sharing workaround for Chrome 88 in FAQs.
 - Add support for Chrome for iOS and Firefox for iOS.
+- Fix Cannot read property 'toLowerCasePropertyNames' of undefined
```
Comment (richnew10) [inline CHANGELOG.md:18]: This needs to move to the 'Unreleased' section.

File: src/meetingsession/MeetingSessionConfiguration.ts [hunk 2]
```diff
@@ -173,7 +173,11 @@ export default class MeetingSessionConfiguration {
     } else if (typeof input !== 'object') {
       return input;
     } else if (Array.isArray(input)) {
-      return input.map(this.toLowerCasePropertyNames);
+      if (input.length) {
+        return input.map(this.toLowerCasePropertyNames);
+      } else {
+        return input;
+      }
```
Comment (richnew10) [inline src/meetingsession/MeetingSessionConfiguration.ts:180]: I'm not entirely sure what you're trying to fix here, but I suspect the correct thing to do is to move `toLowerCasePropertyNames` out as a free function. Even with your change this use of `this` is incorrect.

**Review by richnew10 — CHANGES_REQUESTED** [review]
See inline.

**Comment by nainkunal933** [comment]
> > > @ltrung I fixed the unit tests. The CI fails and I am not sure why. Can you please review it?
> > 
> > 
> > @narek-king currently forked PR would not be able to trigger our integration test as it cannot access GitHub secret. We are working on fixing it.
> > Regarding your PR, could you clarify what the issue is? I could run the unit test you added successfully without the fix. [].map should not cause any issue.
> 
> @ltrung I agree that [].map should not cause any issue but I have a real case that it didn't work. This fixed my problem and I am currently using the SDK from my forked repo. In my opinion, the problem is not related to [].map but more of recursively calling the `toLowerCasePropertyName` method. In the scope, the reference of the method is not accessible. I was thinking to refactor to call as a static method but in this case, it doesn't make sense.
> I was thinking of another case where the input is [[]] but didn't have time to check.
> image
> I am not sure why this is hapenning^

Hi @narek-king, thank you for opening this PR. I have looked through your code fix. It seems like you are checking for array length before calling `return input.map(this.toLowerCasePropertyNames);`

The reason why the code is throwing undefined error here is because of `Array.map()`'s behavior. Here is quote from the methods implementation.

>“If a thisArg parameter is provided, it will be used as callback’s this value. Otherwise, the value undefined will be used as its this value. The this value ultimately observable by callbackFn is determined according to the usual rules for determining the this seen by a function.”

If I understand correctly, this is why @richnew10 is referring to the usage of `this` as incorrect in the [comment above](https://github.com/aws/amazon-chime-sdk-js/pull/1072#discussion_r579714994).

One way to fix this will be to pass `this` as a param to the map method, like this:
Screen Shot 2021-09-09 at 12 05 25 PM

Please refer to [MDN docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map) for more info on this optional parameter.

I will be raising a PR soon to fix this by making the method static. I have updated the issue #1052 with my comment. Please let me know if you have any questions.

**Comment by nainkunal933** [comment]
> @ltrung I fixed the unit tests. The CI fails and I am not sure why. Can you please review it?

@narek-king The CI is failing because external PRs that are coming from a fork do not have access to GitHub secrets. So, you will see that the integration tests are not running. Currently, we do not have a fix for this issue and we plan to resolve this in the future.

Because the tests will not run we will not be able to merge this PR. I am sorry for this inconvenience. I want to thank you for taking the time to open this PR. Your PR was very helpful in understanding the issue. Please refer to the issue #1052 as I will keep you updated there and feel free to review my PR 😄

**Pull request closed without merging**